packages/core/src/json.ts

```diff
@@ -0,0 +1,115 @@
+import { StateNode, ActionObject, Guard, InvokeDefinition } from './';
```

Comment:
This file is currently unused

Reply:
What problems have you encountered with this? Is the plan to use this file in the future?

Reply:
Yes, I plan to use this file in the future for JSON parsing utilities and integration with other XState tools.

Reply:
I'm wondering if this will be the future way of serializing machines etc and if yes - then we could avoid adding `toJSON` to formatted transitions because we could just move its logic to a `machineToJson` or alike

Reply:
I still want `toJSON` to work as-is.

This file will be the basis for serializing machine definitions including non-serializable things such as functions, maps, sets, etc. and it be interpreted correctly.

Reply:
I personally would prefer to keep this out - serializing machines is an additional feature that doesn't have to belong to the "core" core, especially that most people won't really use it that way. While XState allows serialization, and this is great, it's just an added benefit - not something that is actually much needed for most runtime apps.

I understand if you feel differently and won't argue about it much. Just I prefer dedicated functions for stuff, rather than baking too much into the core.

Reply:
Especially that there is no builtin to actually rehydrate jsonified stuff in JavaScript, so this:
```js
jsonToMachine(machineToJson(machine))
```
makes the parity between those nicer than:
```js
jsonToMachine(JSON.stringify(machine))
```

Reply:
The assumption is that developers using `JSON.stringify/parse` should have already "stringified" their implementation details. I want to make sure that that behavior works.

In fact, `machineToJson` etc. should _not_ be the common use-case, because it will essentially be `eval()`-ing code such as functions defined as strings, which can be risky.

Reply:
That is, given the developer already has the implementation details of actions, service, etc., it should be possible to `JSON.parse(someMachineString)` and apply it as such:

```js
const machineConfig = JSON.parse(someMachineString);

// Implementation details are not part of the machine config!
const machine = createMachine(machineConfig, {
  actions: /* ... */,
  services: /* ... */,
  guards: /* ... */,
  // ... etc.
});
```

This is the main use-case I want to support.

Reply:
Right, this is IMHO OK - this is orthogonal to `JSON.stringify` though. Do you suspect having to call smth like `stringify(machineConfig)` over `JSON.stringify` would deteriorate DX that much? Stringifying seems like a use case for rather advanced users and as such, I wouldn't expect it to be a problem.

I know this is somewhat in opposite to what I have raised previously about not having the parity between serialize & deserialize (where `JSON.parse` serves as deserializing algorithm here), but I wasn't sure how exactly you have envisioned this to be used. 

I still think it, in general sense, would be great to have the parity between those 2. So maybe instead of adding custom `toJSON` to some internal structure, we should think of how to just represent the whole definition in a serializable way. We can split the serializable definition from the other data about the transition that we need at runtime (or maybe we wouldn't need splitting it - just a different representation).

If you don't want to introduce a separate stringifier then this is OK to be merged. We can try to pursue different definition representation in other PRs.

Also - `Object.defineProperty` is not the fastest thing in the world. Or at least it hasn't been in the past: https://humanwhocodes.com/blog/2015/11/performance-implication-object-defineproperty/ . Browsers perf characteristics are a moving target and it would have to be checked if this is still the case, but if this is uncertain - it's better to avoid it altogether (if possible ofc).

Reply:
> ```js
> const machineConfig = JSON.parse(someMachineString);
> 
> // Implementation details are not part of the machine config!
> const machine = createMachine(machineConfig, {
>   actions: /* ... */,
>   services: /* ... */,
>   guards: /* ... */,
>   // ... etc.
> });
> ```
> 
> This is the main use-case I want to support.

Just as the feedback this exactly one I'm currently doing it. 
I have a `createMyMachine` which given some business objects will return an xState plain `machineConfig` that can be used later. 
I also have the serialization scenario so as you said I need to handle `"stringified" their implementation details`

So having the JSON schema plus some utils to deal with that is pretty cool 🙏